### ISSUE TITLE
Use service.load(class) instead of service.load(class, classloader) for OSGi compat

### ISSUE THREAD

**Issue by HuanLinying** [grpc/grpc-java#3273]
Please answer these questions before submitting your issue.

### What version of gRPC are you using?
1.4.0

### What JVM are you using (`java -version`)?
1.8
****
### What did you do?
I'm trying to wrap GRPC jar to OSGI bundle. There is a problem of making java.util.service works in OSGI environment. The reason is GRPC use service.load api with classloader parameter, such as:

    @VisibleForTesting
    public static Iterable<ManagedChannelProvider> getCandidatesViaServiceLoader(
      ClassLoader classLoader) {
    return _ServiceLoader.load(ManagedChannelProvider.class,_ classLoader);
    }

I‘m using Apache Aries spi-fly to transferring  java.util.service to OSGI service. But unfortunately this plugin does not support the api service.load(class, classloader) now, and there is no other way to figure this issue out. Actually assigning  classloader is conflicts with OSGI in my opinion.

### What did you expect to see?
It seems no necessary to assign classloader parameter when use service.load(), could we just use service.load(class)?

**Comment by ejona86** [comment]
Since #3183 (which didn't make it in 1.5), the argument is necessary, because we use a different class loader (getClass().getClassLoader) than the default (context class loader). It seems like that SPI Fly will be unnecessary in 1.6?

Are you needing SPI Fly for anything else, or only gRPC? If only for gRPC, then it seems you should just stop using SPI Fly in 1.6.

**Comment by HuanLinying** [comment]
> Since #3183 (which didn't make it in 1.5), the argument is necessary, because we use a different class loader (getClass().getClassLoader) than the default (context class loader). It seems like that SPI Fly will be unnecessary in 1.6?
> 
> Are you needing SPI Fly for anything else, or only gRPC? If only for gRPC, then it seems you should just stop using SPI Fly in 1.6.

It sounds great. I'm gonna check it out and feed back soon.

**Comment by ejona86** [comment]
We have no knowledge with OSGi, so it is hard to help.

> gRPC-core and gRPC-netty are separate OSGI bundles

Can you just put them in the same bundle?

Taking a quick glance at [SPI Fly's documentation about OSGi-specific config](http://aries.apache.org/modules/spi-fly.html#specconf), it seems I'd have to understand OSGi considerably more, because it seems incomplete at face value (how does that system work if there are two copies of gRPC at runtime? Maybe the answer is Provide-Capability). It would take a lot more investigation for me to understand what is going on, how OSGi expects to work, and figure out a way to make OSGi work without breaking others.

**Comment by HuanLinying** [comment]
Yes, after change classloader to` ManagedChannelProvider.class.getClassLoader())`, I think it will work by putting all gRPC jars to a same OSGI bundle. But separate bundles looks be a better way :)

**Comment by pieterjanpintens** [comment]
Would it be possible to use the load without the extra classloader as fallback if no providers are found with the load with classloader. This way osgi can work and your jars can be converted to separate osgi bundles instead of merging them to one big jar file for osgi.

**Comment by ejona86** [comment]
@pieterjanpintens, that... could be an option. I think it could cause some confusion outside of OSGi, but only when already broken. So that confusion may be acceptable.

What is the simplest OSGi environment to set up/test with? It'd be great to have, say, the helloworld example tweaked to use OSGi.

**Comment by pieterjanpintens** [comment]
@ejona86 I will check if after this modification I can get further with OSGi. If I can get it to work i will post back. It can take a couple of days because I have only limited time for this research project on grpc. I it all works out I can publish the code on github. That is code to covert the required plain jars to OSGi (that is grcp and some of its dependencies) and a little sample server / client.

**Comment by pieterjanpintens** [comment]
I was able to get most of it working. There is still a little workarround in the server initialization code on my side. I added an extra dependency on my class arround the GRCP server instance that waits for ServerProvider service to come live. Otherwise the static initialization of the server itself would not see the new providers. This feels like a typical OSGi thing where services can come and go but since load is only called once the services must already be there (or we must call load more).
Anyway my demo code is here https://github.com/pieterjanpintens/grpc-osgi.

I tried to run the test after my fix in your code but now two test fail. The reason why is not very clear to me.
The test that fail are those using Iterable<?> NameResolverProviderTest::invokeGetCandidatesViaHardCoded(ClassLoader cl)

They fail with:
java.util.ServiceConfigurationError: io.grpc.NameResolverProvider: Provider io.grpc.internal.DnsNameResolverProvider not a subtype

Which happens before any of the test code but in the static load when the NameResolverProvider class calls the static load method. I need to dig deeper how this comes because they seem to be Class.assignable from the first point of view. Might be because of classloader hussle in the test code?

I will create a pull request asap.
